### ISSUE TITLE
RINA

### ISSUE THREAD

**Issue by dvc94ch** [libp2p/specs#144]
Hi,

Wanted to ask if you are aware of RINA and what it's relationship is to libp2p. Here is my take on it:

RINA is a blueprint for a layer in a networking stack. It's function is configured by a set of policies. It consists of two wire protocols (EFCP and CDAP) and a programming model. There are RINA shims for Ethernet, 802.11 and IP/UDP/TCP. [0] [1] [2] [3]

Libp2p is a virtual overlay network on top of a physical network (TCP/IP), with a bunch of new or existing protocols for communication and management of the layer.

I wonder if it makes sense to adapt the RINA model of two protocols (EFCP and CDAP), make what is currently a protocol in libp2p a selectable policy for those protocols and adopt the RINA programming interface for libp2p applications.

- [0] Patterns in Network Architecture by John Day
- [1] http://nes.fit.vutbr.cz/ivesely/specs/pmwiki.php/RINA/RINA
- [2] https://github.com/rlite/rlite
- [3] https://github.com/IRATI/stack

**Comment by DerickEddington** [comment]
I'd also like to know whether libp2p is/can-be/will-be compatible with RINA (and other similar recursive designs like [Ouroboros](https://ouroboros.ilabt.imec.be/)).  I recently put some time into researching it and it's very impressive and attractive, and convincingly claims to fix many outstanding and growing problems with IP networks, and has a substantial community of experts interested in furthering it.

I'm now dreaming of future unstoppable mesh networking building on it because of its advantages over IP and its clean design.  It claims advantages in addressing, security, multi-homing, mobility, QoS, NAT, and more.  It'd be very sad if libp2p somehow ended up being incompatible with it, because libp2p is so awesome and well designed, and this has now become a criteria for my interest in libp2p's long-term future.

My initial guesstimate is that libp2p's current design probably is mostly, perhaps fully, compatible with RINA since [non-IP networks](https://github.com/libp2p/specs/blob/master/7-properties.md#74-non-ip-networks) have already been taken into consideration and since [the addressing design is agnostic](https://github.com/libp2p/specs/blob/master/3-requirements.md).  Unfortunately, I haven't had the time to become familiar enough with libp2p and RINA to really verify their compatibility.  (I'd love to be able to explore this and am looking for a new job if anyone wants to pay me.  I'd want to focus on Rust implementations because I'm intent on working with it.)

In any case, RINA is definitely something that the libp2p community should be familiar with and have an explanation for why it is or isn't compatible, IMHO.

Some RINA proponents strongly criticize the founders of the Internet, which might offend some people, but please keep in mind that history is often multi-sided, unfair, ugly, and confusing, and we need to respect valid disagreements, and some of the criticizers are likewise founders even if they weren't given celebrity status.

This is a good intro: https://en.wikipedia.org/wiki/Recursive_Internetwork_Architecture, and you can find many other articles and videos about it.

**Comment by dvc94ch** [comment]
It doesn't help that the book patterns in network architecture is long winded, not to the point and often very focused on history and personal opinion. The book feels like he wrote it from start to finish without any editing or proof reading. It took me quite a while - due to imo poor writing to decide if there was anything new or interesting in RINA.

**Comment by DerickEddington** [comment]
I haven't read Day's book but I did find his oral presentations (videos) to be hard to follow and I did have to read different descriptions over many hours from others and take them at their word before feeling like I was starting to grok it.  Teaching isn't his specialty apparently, though I do value his historical perspectives (which are probably biased to an extent like most humans).

What RINA claims is deep and broad so difficulty presenting it without massive effort and funding seems understandable, and I totally agree it needs better material.  I certainly don't understand it very well presently but it seems worth being given its chance to properly try out.
